middle_end/flambda2/simplify/simplify_let_expr.ml

```diff
@@ -222,66 +222,74 @@ let update_data_flow dacc closure_info ~lifted_constants_from_defining_expr
 let simplify_let0 ~simplify_expr ~simplify_toplevel dacc let_expr ~down_to_up
     bound_pattern ~body =
   let module L = Flambda.Let in
+  let original_dacc = dacc in
   (* Remember then clear the lifted constants memory in [DA] so we can easily
      find out which constants are generated during simplification of the
      defining expression and the [body]. *)
   let dacc, prior_lifted_constants = DA.get_and_clear_lifted_constants dacc in
   (* Simplify the defining expression. *)
-  (* CR mshinwell/vlaviron: We shouldn't need to continue simplifying the body
-     if [Simplify_named] returns [Invalid]. *)
   let simplify_named_result, removed_operations =
     Simplify_named.simplify_named dacc bound_pattern (L.defining_expr let_expr)
       ~simplify_toplevel
   in
-  let dacc = Simplify_named_result.dacc simplify_named_result in
-  (* First accumulate variable, symbol and code ID usage information. *)
-  (* CR-someday gbury/pchambart : in the case of an invalid, we currently
-     over-approximate the uses. In case of an invalid, we might want to instead
-     flush the uses of the current control flow branch (but this would require a
-     more precise stack). *)
-  (* We currently over-approximate the use of variables in symbols: both in the
-     lifted constants, and in the bound constants, which we consider to be
-     always used, leading to the free_names in their defining expressions to be
-     considered as used unconditionally. *)
-  let closure_info = DE.closure_info (DA.denv dacc) in
-  (* Next remember any lifted constants that were generated during the
-     simplification of the defining expression and sort them, since they may be
-     mutually recursive. Then add back in to [dacc] the [prior_lifted_constants]
-     remembered above. This results in the definitions and types for all these
-     constants being available at a subsequent [Let_cont]. At such a point,
-     [dacc] will be queried to retrieve all of the constants, which are then
-     manually transferred into the computed [dacc] at the join point for
-     subsequent simplification of the continuation handler(s).
+  (* We don't need to simplify the body of the [Let] if the defining expression
+     simplified to [Invalid]. *)
+  if Simplify_named_result.is_invalid simplify_named_result
+  then
+    down_to_up original_dacc ~rebuild:(fun uacc ~after_rebuild ->
+        let uacc = UA.notify_removed ~operation:removed_operations uacc in
```

Comment:
We're removing the body of the let binding, and it won't be taken into account here. Is this fine ?

Reply:
Not ideal, but I'm unsure how to compute what has been removed (it seems like we should try to avoid traversing the invalid code if possible).  @poechsel maybe you have an idea?

Reply:
ping @poechsel ^

Reply:
I don't have any good idea on how to do that. One solution could be to compute the code size during closure conversion and store it in the let expression but I don't know how well would it work with lifting...

Reply:
I think maybe we should just accept we can't adjust the benefit in this case.  Doesn't seem like a showstopper.

Reply:
So let's merge this PR and later see if we need to change things to better account the benefits in this case ?